modules/ingest-geoip/src/main/java/org/elasticsearch/ingest/geoip/DatabaseRegistry.java

```diff
@@ -0,0 +1,330 @@
+                try {
+                    Files.delete(databaseTmpFile);
+                    Files.delete(databaseTmpGzFile);
+                } catch (IOException ioe) {
+                    ioe.addSuppressed(failure);
+                    LOGGER.error("Unable to delete tmp database file after failure", ioe);
+                }
+            });
+    }
+
+    void updateDatabase(String databaseFileName, String recordedMd5, Path file) {
+        try {
+            LOGGER.info("database file changed [{}], reload database...", file);
+            DatabaseReaderLazyLoader loader = new DatabaseReaderLazyLoader(cache, file, recordedMd5);
+            DatabaseReaderLazyLoader existing = databases.put(databaseFileName, loader);
+            if (existing != null) {
+                existing.close();
+            }
+        } catch (Exception e) {
+            LOGGER.error((Supplier<?>) () -> new ParameterizedMessage("failed to update database [{}]", databaseFileName), e);
+        }
+    }
+
+    void removeStaleEntries(Collection<String> staleEntries) {
+        for (String staleEntry : staleEntries) {
+            try {
+                LOGGER.info("database [{}] no longer exists, cleaning up...", staleEntry);
+                DatabaseReaderLazyLoader existing = databases.remove(staleEntry);
+                assert existing != null;
+                existing.close(true);
+            } catch (Exception e) {
+                LOGGER.error((Supplier<?>) () -> new ParameterizedMessage("failed to clean database [{}]", staleEntry), e);
+            }
+        }
+    }
+
+    void retrieveDatabase(String databaseName,
+                          String expectedMd5,
+                          GeoIpTaskState.Metadata metadata,
+                          CheckedConsumer<byte[], IOException> chunkConsumer,
+                          CheckedRunnable<Exception> completedHandler,
+                          Consumer<Exception> failureHandler) {
+        // Need to run the search from a different thread, since this is executed from cluster state applier thread:
+        genericExecutor.accept(() -> {
+            MessageDigest md = MessageDigests.md5();
+            Long lastSortValue = null;
+
+            try {
+                do {
+                    SearchRequest searchRequest =
+                        createSearchRequest(databaseName, metadata.getFirstChunk(), metadata.getLastChunk(), lastSortValue);
+                    lastSortValue = null;
+
+                    // At most once a day a few searches may be executed to fetch the new files,
+                    // so it is ok if this happens in a blocking manner on a thead from generic thread pool.
+                    // This makes the code easier to understand and maintain.
+                    SearchResponse searchResponse = client.search(searchRequest).actionGet();
+                    if (searchResponse.getHits().getHits().length == 0) {
```

Comment:
Another idea: we know exactly which chunks we need by id we can do
```java
for(int i = firstChunk; i <= lastChunk; i++) {
   Map<String, Object> source = client.prepareGet(DATABASES_INDEX, name + "_" + i).get().getSource();
   ...
}
```
It should be simpler than search with last sort value

Reply:
I like that idea. This should make this code here much easier. 
Also a search by id would be fine too.

Reply:
I pushed: acf2a4b2a322871caaac7b96f6a08a311e6885bd